input-profile.js

```diff
@@ -0,0 +1,249 @@
+        }
+    }
+
+    update() {
+        if (this.modelLoaded) {
+            this.mapGamepadInput();
+        }
+    }
+
+    onInputSourcesChange(event) {
+        event.added.forEach((xrInputSource) => {
+            if (xrInputSource.hand != undefined) {
+                this.handleHandSource();
+                return;
+            } else {
+                this.handleControllerSource(xrInputSource);
+            }
+        });
+    }
+
+    handleControllerSource(xrInputSource) {
+        this.activatePrespawnedControllersOnStart();
+        this.loadProfileJson(xrInputSource);
+    }
+
+    handleHandSource(xrInputSource) {
+        this.deactivatePreSpawnedControllerOnStart();
+    }
+
+    activatePrespawnedControllersOnStart() {
+        this.setComponentsActive(true);
+    }
+
+    deactivatePreSpawnedControllerOnStart() {
+        this.setComponentsActive(false);
+    }
+
+    getComponents(obj) {
+        const comps = obj.getComponents().filter((c) => c.type !== 'vr-mode-active-switch');
+        this.components = this.components.concat(comps);
+
+        let children = obj.children;
+        for (let i = 0; i < children.length; ++i) {
+            this.getComponents(children[i]);
+        }
+    }
+
+    setComponentsActive(active) {
+        const comps = this.components;
+        for (let i = 0; i < comps.length; ++i) {
+            comps[i].active = active;
+        }
+    }
+
+    loadProfileJson(xrInputSource) {
+        /** load Profile json */
+        /** Note: if you are providing a custom ProfileJSON provide its path in the
+         * customProfileFolder property of the input-profile component.
+         */
+        this.profile = this.customProfileFolder || this.path + xrInputSource.profiles[0];
+        this.url = this.profile + '/profile.json';
+        if (Object.keys(this.ProfileJSON).length > 0) {
+            this.loadModel(xrInputSource);
+        } else {
+            fetch(this.url)
+                .then((res) => res.json())
+                .then((out) => {
+                    this.ProfileJSON = out;
+                    this._profileJsonLoaded = true;
+                    this.loadModel(xrInputSource);
+                })
+                .catch((error) => {
+                    console.log('couldnot read profile.json ' + this.url);
+                });
+        }
+    }
+
+    loadModel(xrInputSource) {
+        if (this.handedness === xrInputSource.handedness) {
+            this.gamepad = xrInputSource.gamepad;
+            /** load controllerModel **/
+
+            if (this.modelLoaded) return;
+
+            if (this.preSpawnedController !== null) {
+                this.controllerObject = this.preSpawnedController;
+                this.controllerObject.parent = this.object;
+                this.controllerObject.setTranslationLocal([0, 0, 0]);
+                this.getGamepadObjectsFromProfile(this.ProfileJSON, this.controllerObject);
+                this.modelLoaded = true;
```

Comment:
Instead of using a `modelLoaded` boolean, you could just initialize `controllerObject` to `null`, and check if that's null? You can even create a getter:


```js
get isModelLoaded() { return this.controllerObject !== null; }
```
```